File: crates/primitives/src/js/column.ts
import { DataTypes } from "ext:sjs_primitives/src/js/dataTypes.ts";

export class Column {
    public name: string;
    public dataType: DataTypes;
    public defaultValue?: string;
    public comment?: string;
    public required: boolean = false;
    public primaryKey: boolean = false;

    constructor(name: string, dataType?: DataTypes) {
        this.name = name;
        this.dataType = dataType || DataTypes.String;
    }

    string() {
        this.dataType = DataTypes.String;
        return this;
    }

    boolean() {
        this.dataType = DataTypes.Boolean;
        return this;
    }

    require(data: boolean) {
        this.required = data;
        return this;
    }

    withComment(comment: string) {
        this.comment = comment;
        return this;
    }

    withDefaultValue(val: any) {
        const mapping = {
            [DataTypes.String]: {
                type: 'string',
                validator: (x:any) => typeof x === 'string'
            },
            [DataTypes.Boolean]: {
                type: 'boolean',
                validator: (x:any) => typeof x === 'boolean'
            }
        };

        const mapEntry = mapping[this.dataType];
        if (mapEntry && mapEntry.validator && !mapEntry.validator(val)) {
            throw new Error(`Default value does not match column type. ${this.name} is of type '${mapEntry.type}'.`);
        }

        this.defaultValue = String(val);
        return this;
    }

}

File: crates/primitives/src/column.rs
use crate::types::DataTypes;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Column {
    pub name: String,
    pub data_type: DataTypes,
    pub default_value: Option<String>,
    pub required: bool,
    pub comment: Option<String>,
}

impl Column {
    pub fn new(name: &str, data_type: DataTypes) -> Self {
        Self {
            name: String::from(name),
            data_type,
            default_value: None,
            comment: None,
            required: false,
        }
    }

    pub fn set_default_value(mut self, default_value: &str) -> Self {
        self.default_value = Some(default_value.to_string());
        self
    }

    pub fn set_comment(mut self, comment: &str) -> Self {
        self.comment = Some(comment.to_string());
        self
    }
}
